cabal-validate/src/Main.hs

```diff
@@ -0,0 +1,430 @@
+cabalListBinArgs opts = "list-bin" : cabalArgs opts
+
+-- | Get the binary for a given @cabal@ target by running @cabal list-bin@.
+cabalListBin :: Opts -> String -> IO FilePath
+cabalListBin opts target = do
+  let args = cabalListBinArgs opts ++ [target]
+  stdout <-
+    readProcessStdout_ $
+      proc (cabal opts) args
+
+  pure (T.unpack $ T.strip $ T.toStrict $ T.decodeUtf8 stdout)
+
+-- | Get the RTS arguments for invoking test suites.
+--
+-- These seem to only be used for some of the test suites, I'm not sure why.
+rtsArgs :: Opts -> [String]
+rtsArgs opts =
+  case archPath opts of
+    "x86_64-windows" ->
+      -- See: https://github.com/haskell/cabal/issues/9571
+      if compilerVersion (compiler opts) > makeVersion [9, 0, 2]
+        then ["+RTS", "--io-manager=native", "-RTS"]
+        else []
+    _ -> []
+
+-- | Run a binary built by @cabal@ and output timing information.
+--
+-- This is used to run many of the test suites.
+timedCabalBin :: Opts -> String -> String -> [String] -> IO ()
+timedCabalBin opts package component args = do
+  command <- cabalListBin opts (package <> ":" <> component)
+  timedWithCwd
+    opts
+    package
+    command
+    args
+
+-- | Print the configuration for CI logs.
+printConfig :: Opts -> IO ()
+printConfig opts = do
+  putStrLn $
+    "compiler:          "
+      <> compilerExecutable (compiler opts)
+      <> "\ncabal-install:     "
+      <> cabal opts
+      <> "\njobs:              "
+      <> show (jobs opts)
+      <> "\nsteps:             "
+      <> unwords (map displayStep (steps opts))
+      <> "\nHackage tests:     "
+      <> show (hackageTests opts)
+      <> "\nverbose:           "
+      <> show (verbose opts)
+      <> "\nextra compilers:   "
+      <> unwords (extraCompilers opts)
+      <> "\nextra RTS options: "
+      <> unwords (rtsArgs opts)
+
+-- | Print the versions of tools being used.
+printToolVersions :: Opts -> IO ()
+printToolVersions opts = do
+  timed opts (compilerExecutable (compiler opts)) ["--version"]
+  timed opts (cabal opts) ["--version"]
+
+  forM_ (extraCompilers opts) $ \compiler' -> do
+    timed opts compiler' ["--version"]
+
+-- | Run the build step.
+build :: Opts -> IO ()
+build opts = do
+  printHeader "build (dry run)"
+  timed
+    opts
+    (cabal opts)
+    ( cabalNewBuildArgs opts
+        ++ targets opts
+        ++ ["--dry-run"]
+    )
+
+  printHeader "build (full build plan; cached and to-be-built dependencies)"
+  timed
+    opts
+    "jq"
+    [ "-r"
+    , -- TODO: Maybe use `cabal-plan`? It's a heavy dependency though...
+      ".\"install-plan\" | map(.\"pkg-name\" + \"-\" + .\"pkg-version\" + \" \" + .\"component-name\") | join(\"\n\")"
+    , baseBuildDir opts </> "cache" </> "plan.json"
+    ]
+
+  printHeader "build (actual build)"
+  timed
+    opts
+    (cabal opts)
+    (cabalNewBuildArgs opts ++ targets opts)
+
+-- | Run doctests.
+--
+-- This doesn't work on my machine, maybe @cabal.nix@ needs some love to
```

Comment:
The doctest step in validate.sh (which is what you were inspired by, I guess) has bit rotted. Look at how we call doctest in the root Makefile. That’s what doctest CI job uses.  Perhaps, that logic should move to validate and the Makefile should only call the corresponding validate step.

Reply:
> The doctest step in validate.sh (which is what you were inspired by, I guess) has bit rotted.

To be clear: Nothing in this PR is 'inspired', this is a 1:1 translation of `validate.sh`, including many silly details. I'd love to work on fixing it up once this PR is merged, but for now I want to keep the `validate.sh` behavior unchanged.